### ISSUE TITLE
askx deploy missing functionality that's in ask: --target in ask deploy

### ISSUE THREAD

**Issue by goedym** [alexa-labs/ask-cli#44]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Other... Please describe:
</code></pre>

## Expected Behavior
I'm trying to deploy an Alexa skill using askx but it's missing the target option. I want to be able to deploy skill and model without deploying lambda.

## Current Behavior
ask cli is able to target lambda, skill, isp, model, or all.  

## Possible Solution
Add full support for all options available in ask cli v1.  (It's likely already on the roadmap.)

**Comment by jameswilsterman** [comment]
I second this feature request.

**Comment by kiernana** [comment]
+1 this request

**Comment by nikhilym** [comment]
Thanks for the feature request. We will take a note of it. Btw it would be great if others wanting this feature can upvote the issue. Thanks!!

**Comment by RonWang** [comment]
Hi @goedym @jameswilsterman @kiernana  (and all who has the same request),

Thanks for the feature request, it is in our roadmap and we'd also like to understand further on what target you really need before we implement.

Before collecting your opinions, I'd think it necessary to compare those two versions of ask-cli for better understanding. As you mentioned, we do have the `--target` option in ask-cli v1 to help skill developers control their deployments. We support **lambda, skill, isp, model, or all** as target values. However, we want to help make the deployment intelligent so there are some differences between the two versions that's related to the usage of `--target`, specifically in latest ask-cli-x:
* We use [Skill Package format](https://developer.amazon.com/en-US/docs/alexa/smapi/skill-package-api-reference.html#skill-package-format) to deploy Alexa skill's metadata, which doesn't support target-deployment yet (see [import-skill-package](https://developer.amazon.com/en-US/docs/alexa/smapi/skill-package-api-reference.html#import-existing-skill)). The v1's skill, model, isp are all the concepts in skill package/metadata deployment, and we might need to forward this feature request to service team.
* We only deploy when necessary by using `lastDeployHash` for skill-package folder and Lambda code folder in ask-cli-x. You might have noticed that we now display a yellow warning message to skip a certain deployment step when nothing changes after comparing with `lastDeployHash`. This is indeed the strategy we use now which should optimize the deployment flow in our design and might save the usage of `--target`. 
* The underlying platform services that we deploy to can also resolve the changesets before deciding if a deployment is needed:
  * Skill infrastructure can be through AWS Cloudformation (if using `@ask-cli/cfn-deployer` as infrastructure deployer) which only deploys the minimal changeset between current and last time deploy.
  * Skill metadata is deployed through Alexa skill package service as mentioned above, which can also support similar experience as AWS Cloudformation does (as a feature request).

And if for your most related question
> I want to be able to deploy skill and model without deploying lambda.

This can be easily achieved by modifying the ask-resources.json file to remove the entire `skillInfrastructure` field. This file controls the deploy targets for ask-cli-x, i.e. skillMetadata, code(build) and skillInfrastructure. If you remove this `skillInfrastructure` step, it will only deploy your skillMetadata and build your code.

___

After all the context, we'd like to understand:
* Do you still think having an explicit `--target` option to control deployment steps in the command interface is needed?
* If still needed:
  * Do you need it more to control the deployment of Alexa versus AWS infrastructures, or the deployment inside Alexa's JSON files?
  * For Alexa's JSON files (e.g. skill, model, isp, or task), do you have the need to deploy one specific resource inside skill package? Can it be fulfilled if skill package supports changeset-based deployment as mentioned above?

A long answer lol but feel free to let me know if any bullet point needs more detailed explanation.

Thanks,
Ron

**Comment by goedym** [comment]
Thanks for the the detailed response/questions.  I don't know all the use cases out there, but I know for me it's sufficient to just bypass the lambda deployment to AWS step.  I'd already figured out how to bypass that myself the day I entered the original ticket by modifying the askx code, probably along the lines of what you described.  (I'd have to go look -- I can't recall what I did.) . I'd not like to be maintaining custom code though otherwise it'll be awkward keeping in sync with askx enhancements.  
So ... to answer your questions:  I
1.  I don't absolutely need a target option per se.  I'd be happy with a switch that disabled lambda deployment.  Or if that could be done with an attribute in ask-resources.json, that would also work for me.  I'm not sure what makes the most sense for you in the bigger picture.
2. I don't need more control of deployment other than that.  But again, there might be other users out there that do.  I know my use case, but not what others might be doing.  Maybe it'd be worth opening these questions up to the broader user base.  i know there were a few others who liked this request, so they might want to provide some input.
